### PR TITLE
[WB-6865] ensure captive key input even if user presses enter

### PR DESCRIPTION
**Pull request by dannygoldstein**
https://wandb.atlassian.net/browse/WB-6865

Description
-----------

Before if someone did `wandb login --relogin` then hit enter without entering anything, the api key prompt would vanish and be a no-op:

```
(wandb-3.7) danielgoldstein@Daniels-MBP ~ % wandb login --relogin
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter: [enter]
(wandb-3.7) danielgoldstein@Daniels-MBP ~ %
```

Now if you hit enter without entering anything, it will reprompt you for an API key, or tell you how to quit out and do a no-op:

```
(wandb-3.7) danielgoldstein@Daniels-MBP ~ % wandb login --relogin
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit: [enter]
wandb: ERROR No API key specified. Please specify an API key
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:
```

Related

Testing
-------

Manually

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------
Ensures API key prompt remains captive when user enters nothing
------------- END RELEASE NOTES --------------------

### CODE REVIEW

**Comment by dannygoldstein** [comment]
added a test to improve coverage

**Comment by vwrj** [comment]
It goes through the "choices" flow again after pressing [enter] instead of the API key line.

**Comment by dannygoldstein** [comment]
@vwrj I think going through the flow in this context again is fine, 

image

it's just a bit confusing given that the error message asks the user to immediately re-specify the API key. I updated the error message to not set that expectation. Aside from that I think the logic / flow of re-prompting for choices at that stage is ok

File: wandb/sdk/wandb_login.py [hunk 1]
```diff
@@ -177,26 +177,27 @@ def update_session(
...
+                continue
+            except TimeoutError:
+                wandb.termlog("W&B disabled due to login timeout.")
+                return None, ApiKeyStatus.DISABLED
+            if key is False:
+                return None, ApiKeyStatus.NOTTY
+            if not key:
+                return None, ApiKeyStatus.OFFLINE
+            else:
+                return key, ApiKeyStatus.VALID
```
Comment (raubitsj) [inline wandb/sdk/wandb_login.py:200]: i think this was better the way it was... else is strange here and not needed

Comment (dannygoldstein) [reply wandb/sdk/wandb_login.py]: removed else

File: wandb/sdk/lib/apikey.py [hunk 2]
```diff
@@ -200,7 +205,7 @@ def write_netrc(host, entity, key):
 
 def write_key(settings, key, api=None, anonymous=False):
     if not key:
-        return
+        raise ValueError("No API key specified.")
```
Comment (raubitsj) [inline wandb/sdk/lib/apikey.py:208]: as long as you have gone through all the paths that could call this and they make sense...  raising a exception seems like a much better thing to do, im just worried there is some path that is calling this with an empty key and expects it to pass

**Review by raubitsj — APPROVED** [review]

File: wandb/sdk/lib/apikey.py (hunk 2 above)
Comment (dannygoldstein) [reply wandb/sdk/lib/apikey.py:208]: all the paths are guarded.  there are 6 in the code:

`configure_api_key()` in `_WandbLogin()` which only is called if `key is not None`, 

and 5 in `prompt_api_key`, which is called only in one place and in that one place has an exception handler (implemented in this PR).

**Pull request merged**